paimon-trino-common/src/main/java/org/apache/paimon/trino/TrinoPageSourceProvider.java

```diff
@@ -64,32 +113,229 @@ public ConnectorPageSource createPageSource(
+                                                                    columnIndex,
+                                                                    rowType.getFields(),
+                                                                    schemaManager
+                                                                            .schema(
+                                                                                    rawFile
+                                                                                            .schemaId())
+                                                                            .fields()),
+                                                            type,
+                                                            orderDomains(projectedFields, filter)))
+                                    .collect(
+                                            Collector.of(
+                                                    LinkedList::new,
+                                                    List::add,
+                                                    (left, right) -> {
+                                                        left.addAll(right);
+                                                        return left;
+                                                    })));
+                } catch (Exception e) {
+                    throw new RuntimeException(e);
+                }
+            } else {
+                // old read way
+                ReadBuilder read = table.newReadBuilder();
+                new TrinoFilterConverter(rowType).convert(filter).ifPresent(read::withFilter);
+
+                if (!fieldNames.equals(projectedFields)) {
+                    read.withProjection(columnIndex);
+                }
+
+                return new TrinoPageSource(
+                        read.newRead().executeFilter().createReader(paimonSplit), columns, limit);
+            }
+        } catch (Exception e) {
+            throw new RuntimeException(e);
         }
+    }
+
+    // make domains(filters) to be ordered by projected fields' order.
+    private List<Domain> orderDomains(
+            List<String> projectedFields, TupleDomain<TrinoColumnHandle> filter) {
+        Optional<Map<TrinoColumnHandle, Domain>> optionalFilter = filter.getDomains();
+        Map<String, Domain> domainMap = new HashMap<>();
+        optionalFilter.ifPresent(
+                trinoColumnHandleDomainMap ->
+                        trinoColumnHandleDomainMap.forEach(
+                                (k, v) -> domainMap.put(k.getColumnName(), v)));
 
-        new TrinoFilterConverter(rowType).convert(filter).ifPresent(read::withFilter);
+        return projectedFields.stream()
+                .map(name -> domainMap.getOrDefault(name, null))
+                .collect(Collectors.toList());
+    }
+
+    private boolean checkRawFile(Optional<List<RawFile>> optionalRawFiles) {
```

Comment:
check like here will throw exception.

You can rename it to `canUseTrinoPageSource`